File: app/jobs/contact_broadcast_job.rb
# frozen_string_literal: true

# Broadcasts individual contact updates for live table refresh
# This job renders the contact row HTML and sends it via ActionCable
class ContactBroadcastJob < ApplicationJob
  queue_as :default

  def perform(contact_id)
    contact = Contact.find_by(id: contact_id)
    return unless contact

    # Broadcast the updated contact data as JSON
    # The JavaScript will update the row in place
    ActionCable.server.broadcast('contacts_updates', {
      action: 'update',
      contact_id: contact.id,
      status: contact.status,
      status_class: contact.status,
      device_type: contact.device_type,
      rpv_status: contact.rpv_status,
      rpv_status_class: rpv_status_class(contact.rpv_status),
      carrier_name: contact.carrier_name,
      risk_level: contact.sms_pumping_risk_level,
      risk_class: risk_class(contact.sms_pumping_risk_level),
      is_business: contact.is_business,
      business_name: contact.business_name,
      formatted_phone: contact.formatted_phone_number
    })
  rescue StandardError => e
    Rails.logger.warn("Contact broadcast failed for #{contact_id}: #{e.message}")
  end

  private

  def rpv_status_class(status)
    return nil unless status

    case status.downcase
    when 'connected' then 'ok'
    when 'disconnected' then 'error'
    else 'warning'
    end
  end

  def risk_class(level)
    return nil unless level

    case level
    when 'high' then 'error'
    when 'medium' then 'warning'
    when 'low' then 'ok'
    end
  end
end

File: laravel-app/app/Jobs/ContactBroadcastJob.php
<?php

namespace App\Jobs;

use App\Models\Contact;
use Illuminate\Bus\Queueable;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Foundation\Bus\Dispatchable;
use Illuminate\Queue\InteractsWithQueue;
use Illuminate\Queue\SerializesModels;
use Illuminate\Support\Facades\Log;
use App\Events\ContactUpdated;
use Exception;

/**
 * Broadcasts individual contact updates for live table refresh
 * This job renders the contact row data and sends it via broadcasting
 */
class ContactBroadcastJob implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;

    public $contactId;

    /**
     * The number of times the job may be attempted.
     *
     * @var int
     */
    public $tries = 1;

    /**
     * The maximum number of seconds the job can run before timing out.
     *
     * @var int
     */
    public $timeout = 10;

    /**
     * Create a new job instance.
     *
     * @param int $contactId
     */
    public function __construct(int $contactId)
    {
        $this->contactId = $contactId;
        $this->onQueue('default');
    }

    /**
     * Execute the job.
     *
     * @return void
     */
    public function handle()
    {
        try {
            $contact = Contact::find($this->contactId);
            if (!$contact) {
                return;
            }

            // Broadcast the updated contact data
            broadcast(new ContactUpdated([
                'action' => 'update',
                'contact_id' => $contact->id,
                'status' => $contact->status,
                'status_class' => $contact->status,
                'device_type' => $contact->device_type,
                'rpv_status' => $contact->rpv_status,
                'rpv_status_class' => $this->rpvStatusClass($contact->rpv_status),
                'carrier_name' => $contact->carrier_name,
                'risk_level' => $contact->sms_pumping_risk_level,
                'risk_class' => $this->riskClass($contact->sms_pumping_risk_level),
                'is_business' => $contact->is_business,
                'business_name' => $contact->business_name,
                'formatted_phone' => $contact->formatted_phone_number,
            ]));

        } catch (Exception $e) {
            Log::warning("Contact broadcast failed for {$this->contactId}: {$e->getMessage()}");
        }
    }

    /**
     * Get RPV status CSS class
     *
     * @param string|null $status
     * @return string|null
     */
    protected function rpvStatusClass($status)
    {
        if (!$status) {
            return null;
        }

        switch (strtolower($status)) {
            case 'connected':
                return 'ok';
            case 'disconnected':
                return 'error';
            default:
                return 'warning';
        }
    }

    /**
     * Get risk level CSS class
     *
     * @param string|null $level
     * @return string|null
     */
    protected function riskClass($level)
    {
        if (!$level) {
            return null;
        }

        switch ($level) {
            case 'high':
                return 'error';
            case 'medium':
                return 'warning';
            case 'low':
                return 'ok';
            default:
                return null;
        }
    }

    /**
     * Handle a job failure.
     *
     * @param  \Throwable  $exception
     * @return void
     */
    public function failed(\Throwable $exception)
    {
        Log::warning("ContactBroadcastJob failed for contact {$this->contactId}: {$exception->getMessage()}");
    }
}
